pkg/resources/custom-smtp/custom-smtp.go

```diff
@@ -0,0 +1,216 @@
+	Password    string
+	Port        string
+	Username    string
+}
+
+func GetCustomAddonValues(serverClient k8sclient.Client, namespace string) (*CustomSmtp, error) {
+
+	addonName := addon.GetName(v1alpha1.InstallationTypeManagedApi)
+	secretName := fmt.Sprintf("addon-%s-parameters", addonName)
+
+	secret := &corev1.Secret{}
+	err := serverClient.Get(context.TODO(), k8sclient.ObjectKey{
+		Name:      secretName,
+		Namespace: namespace,
+	}, secret)
+
+	if err != nil {
+		return nil, err
+	}
+
+	customSmtp := &CustomSmtp{}
+
+	value, ok := secret.Data["custom-smtp-from_address"]
+	if ok {
+		customSmtp.FromAddress = string(value)
+	}
+
+	value, ok = secret.Data["custom-smtp-address"]
+	if ok {
+		customSmtp.Address = string(value)
+	}
+
+	value, ok = secret.Data["custom-smtp-password"]
+	if ok {
+		customSmtp.Password = string(value)
+	}
+
+	value, ok = secret.Data["custom-smtp-port"]
+	if ok {
+		customSmtp.Port = string(value)
+	}
+
+	value, ok = secret.Data["custom-smtp-username"]
+	if ok {
+		customSmtp.Username = string(value)
+	}
+
+	return customSmtp, nil
+}
+
+func ParameterValidation(smtp *CustomSmtp) ValidationResponse {
+
+	valid, partial := false, false
+
+	if smtp.Port != "" {
+		valid = true
+	} else {
+		partial = true
+	}
+
+	if smtp.Address != "" {
+		valid = true
+	} else {
+		partial = true
+	}
+
+	if smtp.Username != "" {
+		valid = true
+	} else {
+		partial = true
+	}
+
+	if smtp.FromAddress != "" {
+		valid = true
+	} else {
+		partial = true
+	}
+
+	if smtp.Password != "" {
+		valid = true
+	} else {
+		partial = true
+	}
+
+	if valid && !partial {
+		return Valid
+	} else if valid && partial {
+		return Partial
+	} else {
+		return Blank
+	}
+
+}
+
+func ParameterErrors(smtp *CustomSmtp) string {
+
+	message := ""
+
+	if smtp.Port == "" {
+		message = message + "Port, "
+	}
+
+	if smtp.Address == "" {
+		message = message + "Address, "
+	}
+
+	if smtp.Username == "" {
+		message = message + "username, "
+	}
+
+	if smtp.FromAddress == "" {
+		message = message + "From_Address, "
+	}
+
+	if smtp.Password == "" {
+		message = message + "Password, "
+	}
+
+	return message
+}
+
+func CreateOrUpdateCustomSMTPSecret(ctx context.Context, serverClient k8sclient.Client, smtp *CustomSmtp, namespace string) (v1alpha1.StatusPhase, error) {
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      CustomSecret,
+			Namespace: namespace,
+		},
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, serverClient, secret, func() error {
+		if smtp == nil {
+			return fmt.Errorf("nill pointer passed for smtp details")
+		}
+
+		if secret.Data == nil {
+			secret.Data = map[string][]byte{}
+		}
+
+		secret.Data["from_address"] = []byte(smtp.FromAddress)
+		secret.Data["host"] = []byte(smtp.Address)
+		secret.Data["password"] = []byte(smtp.Password)
+		secret.Data["port"] = []byte(smtp.Port)
+		secret.Data["username"] = []byte(smtp.Username)
+
+		return nil
+	}); err != nil {
+		return v1alpha1.PhaseInProgress, err
```

Comment:
Should this be PhaseFailed or InProgress? Is there a scenarion where it could be inProgress and then recover?

Reply:
Yap that should be PhaseFailed